src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,591 @@
+        # Verify on the TH that the DUT returns data from the expected attribute path.
+        self.print_step(16, "Read a read request of attribute type list")
+        await self.check_attribute_read_for_type(list)
+
+        # Step 17
+        # TH sends the Read Request Message to the DUT to read an attribute of data type enum.
+        # If the device does not have an attribute of data type enum, skip this step.
+        # Verify on the TH that the DUT returns data from the expected attribute path.
+        self.print_step(17, "Read a read request of attribute type enum")
+        await self.check_attribute_read_for_type(MatterIntEnum)
+
+        # Step 18
+        # TH sends the Read Request Message to the DUT to read an attribute of data type bitmap.
+        # If the device does not have an attribute of data type bitmap, skip this step.
+        self.print_step(18, "Read a read request of attribute type bitmap")
+
+        await self.check_attribute_read_for_type(IntFlag)
+
+        # Step 19
+        # TH sends the Read Request Message to the DUT to read any attribute to an unsupported Endpoint.
+        # DUT responds with the report data action.
+        # Verify on the TH that the DUT sends the status code UNSUPPORTED_ENDPOINT
+
+        self.print_step(19, "Send the Read Request Message to the DUT to read any attribute to an unsupported Endpoint")
+        supported_endpoints = set(self.endpoints.keys())
+        all_endpoints = set(range(max(supported_endpoints)+2))
+        unsupported = list(all_endpoints - supported_endpoints)
+        # Read descriptor
+        result = await self.read_single_attribute_expect_error(endpoint=unsupported[0], cluster=Clusters.Descriptor, attribute=Clusters.Descriptor.Attributes.FeatureMap, error=Status.UnsupportedEndpoint)
+
+        # Step 20
+        # TH sends the Read Request Message to the DUT to read any attribute to an unsupported cluster.
+        # DUT responds with the report data action.
+
+        # Verify on the TH that the DUT sends the status code UNSUPPORTED_CLUSTER
+        self.print_step(20, "Send the Read Request Message to the DUT to read any attribute to an unsupported cluster")
+
+        for endpoint_id, endpoint in self.endpoints.items():
+            for cluster_type, cluster in endpoint.items():
+                if global_attribute_ids.cluster_id_type(cluster_type.id) != global_attribute_ids.ClusterIdType.kStandard:
+                    continue
+
+                all_clusters = set(list(ClusterObjects.ALL_CLUSTERS.keys()))
+                dut_clusters = set(list(x.id for x in endpoint.keys()))
+
+                unsupported = [id for id in list(all_clusters - dut_clusters) if global_attribute_ids.attribute_id_type(id)
+                               == global_attribute_ids.AttributeIdType.kStandardNonGlobal]
+
+                unsupported_attribute = (ClusterObjects.ALL_ATTRIBUTES[unsupported[0]])[0]
```

Comment:
This will throw a key error if the list is empty. You'll need to swap this check and the one below

Reply:
Fixed in newest commit

Reply:
You can also get unlucky here if the cluster doesn't have attributes (ex low power cluster)

Reply:
I tested this to just read LowPower and I didn't see an exception raised.